gui/vm/views.py

```diff
@@ -51,7 +53,32 @@ def home(request):
 
 def start(request, id):
     vm = models.VM.objects.get(id=id)
+    raw_file_cnt = None
+    raw_file_resize = 0
     if request.method == 'POST':
+        if vm.vm_type == 'Container Provider':
+            devices = models.Device.objects.filter(vm__id=vm.id)
+            for device in devices:
+                if device.dtype == 'RAW' and device.attributes.get('boot'):
+                    raw_file_cnt = device.attributes.get('path')
+                    raw_file_resize = device.attributes.get('size')
+
+            with client as c:
+                job_id = c.call('vm.fetch_image', 'RancherOS')
+                status = None
+                while status != 'SUCCESS':
+                    __call = c.call('vm.get_download_status', job_id)
+                    status = __call.get('state')
+                    utils.dump_download_progress(__call)
+                    if status == 'FAILED':
+                        return HttpResponse('Error: failed to download the image!')
+                    elif status == 'ABORTED':
+                        return HttpResponse('Error: download ABORTED!')
+                if status == 'SUCCESS':
+                    prebuilt_image = c.call('vm.image_path', 'RancherOS')
+                    if prebuilt_image and raw_file_cnt:
+                        c.call('vm.decompress_gzip', prebuilt_image, raw_file_cnt)
+                        c.call('vm.raw_resize', raw_file_cnt, raw_file_resize)
```

Comment:
At beginning I made like you are suggesting, I did wrap everything in one big method, however I realize that I can use each method for different purposes inside the VM feature. Also I prefer do it in UNIX style, small methods executing a specific function.

I would prefer keep it as it is.

Reply:
@araujobsd I am not suggesting to merge these three methods, I offer to create fourth method that will call these three in correct order. You see, we're now in process of getting rid of Django GUI and using pure client-side (javascript) GUI that will call middleware methods directly. Exposing details of how Container Provider VM is being deployed to javascript GUI is definitely wrong. That's why I suggest to wrap this into another middleware method and call it from Django GUI here. In another words, the less logic is located inside Django GUI, the better.

Reply:
"Exposing details of how Container Provider VM is being deployed to javascript GUI is definitely wrong." I don't know where it will be definitely  wrong if they are exposed in the API documentation already!

I agree about less complexity at Django side.

So, if I wrap it in another middlewared method I will have another method just to deal with less 'logic' at Django side, as I can't see any problem to have these 3 different methods exposed for the new UI.

Mostly of vm plugins is done in that way. And I still can't see any benefit to wrap it in another method.

Reply:
@araujobsd  @themylogin raises a good point. I think it is overly complex for UI developers and/or API users to have to use these methods. I see no problem in exposing these methods if you like, but they should probably be private.

As a API user I just want to `setup` a new rancheros container, I dont want to `fetch_image`, `setup_sharefs`, `decompress_gzip` and `raw_resize` before.

The less implementation details that are exposed, the better.

Can you tell us a good reason these should be exposed as-is? I guess thats what we are failing to see

Reply:
Yes, I can give good reasons.
1) setup_sharefs: I planned to have in the new UI a place where users will choose where to setup the sharefs, pretty much like as we have nowadays in Jails section.
2) fetch_image: Users in the future can download different prebuilt images for different systems.
3) raw_resize: Users in the future can resize their RAW disk, this method will be used for normal VMs too not only for this container.
4) decompress_gzip: I want make a feature where users can move VM's between two different machines and probably I will use it, although compress_gzip wasn't implemented yet. 

So there is a reason why I made it in this way as I have stated before, I can wrap those methods into a new method, however, I want use those methods in the future.

Also it is good for debug and maintenance to decouple these methods instead of have them all wrapped together.

Reply:
We are not asking to completely remove these methods they would keep working without any change. We just dont think that complexity should rest on UI developers.

Thats our opinion, you dont have to agree with it. Its your call.